cmd/habitat/community/consensus/cluster/cluster.go

```diff
@@ -0,0 +1,110 @@
+package cluster
+
+import (
+	"errors"
+	"io/ioutil"
+	"net/url"
+	"os"
+
+	"github.com/eagraf/habitat/cmd/habitat/community/consensus/raft"
+	"github.com/eagraf/habitat/cmd/habitat/proxy"
+	"github.com/eagraf/habitat/pkg/compass"
+	"github.com/rs/zerolog/log"
+)
+
+type ClusterService interface {
+	Start() error
+
+	CreateCluster(communityID string) error
+	RemoveCluster(communityID string) error
+	JoinCluster(communityID string, address string) error
+	RestoreNode(communityID string) error
+
+	ProposeTransition(communityID string, transition []byte) error
+	GetState(communityID string) ([]byte, error)
+
+	AddNode(communityID string, nodeID string, address string) error
+	RemoveNode(communityID string, nodeID string) error
+}
+
+// ClusterManager is a layer of abstraction that allows multiple Cluster Services to share
+// a common service.
+// TODO Right now, the only cluster service implementation is Raft, so the switching implementation
+// needs to be implemented in the future.
+type ClusterManager struct {
+	raftClusterService *raft.ClusterService
+}
+
+func NewClusterManager() *ClusterManager {
+	raft := raft.NewClusterService()
+	return &ClusterManager{
+		raftClusterService: raft,
+	}
+}
+
+func (cm *ClusterManager) Start(proxyRules *proxy.RuleSet) error {
+	// TODO fix this when centralized port management is implemented
+	url, err := url.Parse("http://0.0.0.0:6000/raft/msg")
+	if err != nil {
+		return err
+	}
+
+	// TODO switch between implementations of cluster services
+	proxyRules.Add("raft-service", &proxy.RedirectRule{
+		Matcher:         "/raft/msg",
+		ForwardLocation: url,
+	})
+
+	err = cm.raftClusterService.Start()
+	if err != nil {
+		return err
+	}
+
+	// Restart any existing communities
```

Comment:
why are we restarting existing communities? I thought we have a different consensus service per community, and it just shares a port for receiving / sending messages

Reply:
Yeah thats correct. They are all listening on the same port through the cluster service (6000 for raft), but each community has its own thread that listens independently and gets messages forwarded to it by the service. Start() gets called whenever habitat is started, and this just tells each community to resume listening and grab all updates to state since the last time habitat went down

Reply:
ohh I see. I'm wondering if we distinguish between "running habitat" and "communities connect to their network". At least in my PR that's the approach I took (you have to explicitly "connect your community"). Removing this difference is probably better UI wise, but gives the user less control. We could also change this in the future.

Reply:
Yeah I think the raft/consensus stuff should eventually do that. There should be some file that keeps track of the nodes current state, in terms of communities that it is connected to, that we keep as a file so we can reconnect only to communities that were in use earlier. Right now, the main goal of this bit of code is just to ensure that restarting does work when we need it